packages/villus/src/utils/common.ts

```diff
@@ -16,13 +17,17 @@ export function toWatchableSource<T = any>(value: Ref<T> | Record<string, any>):
 
 // Uses same component provide as its own injections
 // Due to changes in https://github.com/vuejs/vue-next/pull/2424
+// todo: maybe better to modify func name and add manual client args here
 export function injectWithSelf<T>(symbol: InjectionKey<T>, onMissing: () => Error): T {
   const vm = getCurrentInstance() as any;
+  let client = vm && inject(symbol, vm?.provides?.[symbol as any]);
 
-  const injection = inject(symbol, vm?.provides?.[symbol as any]);
-  if (injection === null || injection === undefined) {
+  if (client) setActiveClient(client);
+  client = activeClient;
+
+  if (client === null || client === undefined) {
     throw onMissing();
```

Comment:
Building on top of the previous comment, we can skip passing the `onMissing` callback as well since it is the same in all usages of this.

```suggestion
    throw new Error('Cannot detect villus Client, did you forget to call `useClient`?');
```